rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,417 @@
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction, EVMTxHash, Timestamp
+from rotkehlchen.utils.misc import bytes_to_address
+
+from .constants import CPT_RAINBOW_SWAPS
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class RainbowCommonDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: ChecksumEvmAddress,
+    ) -> None:
+        super().__init__(evm_inquirer, base_tools, msg_aggregator)
+        self.router_address = router_address
+
+    def _enricher_event(
+            self,
+            event: 'EvmEvent',
+            sender: ChecksumEvmAddress,
+    ) -> None:
+        """Set the event to rainbow event
+        """
+        if (
+            event.location_label != sender or
+            event.address != self.router_address or
+            event.counterparty is not None
+        ):
+            return
+        elif event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE:  # noqa: E501
+            # Find the receive event
+            event.event_type = HistoryEventType.TRADE
+            event.event_subtype = HistoryEventSubType.RECEIVE
+            event.counterparty = CPT_RAINBOW_SWAPS
+            event.notes = f'Receive {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap'  # noqa: E501
+        elif event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE:  # noqa: E501
+            # Find the spend event
+            event.event_type = HistoryEventType.TRADE
+            event.event_subtype = HistoryEventSubType.SPEND
+            event.notes = f'Swap {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in Rainbow Swap'  # noqa: E501
+            event.counterparty = CPT_RAINBOW_SWAPS
+            event.address = self.router_address
+
+    def _find_event_by_log(
+            self,
+            logs: list['EvmTxReceiptLog'],
+            event_signature: bytes,
+    ) -> EvmTxReceiptLog | None:
+        """
+        Find the log by the event signature.
+
+        Args:
+            logs (list[EvmTxReceiptLog]): List of logs.
+            event_signature (bytes): Event signature to be searched.
+
+        Returns:
+            Optional[EvmTxReceiptLog]: Log if found, else None.
+        """
```

Comment:
```suggestion
        """
        Find the log that matches  the provided event signature.

        returns the log event if found or None otherwise.
        """
```

btw what happens if there are multiple wraps? I believe the logic can fail there, do you have anything else like the address to match against?

Reply:
It get the first log, I check the some transactions the transactions and only have one wrap event, theoretically we will only have this event per swap. It's not possible to check by address because this log has the same value as the "target" that is present in input_data and since we cannot use it, there is no way to check since this address is dynamic.